File: deepgoal/core/types/engine.py
from __future__ import annotations

from collections.abc import AsyncIterator
from typing import Literal, Protocol

from pydantic import BaseModel, ConfigDict, Field

from .common import JsonValue
from .enums import PermissionMode
from .hooks import HookEvent, HookHandler
from .tools import McpServerConfig, SkillConfig, ToolConfig


class EngineOptions(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    tools: list[ToolConfig] = Field(default_factory=list)
    mcp_servers: dict[str, McpServerConfig] = Field(default_factory=dict)
    skills: list[SkillConfig] = Field(default_factory=list)
    hooks: dict[HookEvent, list[HookHandler]] = Field(default_factory=dict)
    permission_mode: PermissionMode = PermissionMode.DEFAULT
    max_turns: int | None = None
    system_prompt: str | None = None


class AgentRunRequest(BaseModel):
    prompt: str
    cwd: str
    metadata: dict[str, JsonValue] = Field(default_factory=dict)


class AgentStreamEvent(BaseModel):
    type: Literal["message", "tool_call", "tool_result", "error", "done"]
    content: str | None = None
    metadata: dict[str, JsonValue] = Field(default_factory=dict)


class EngineRunResult(BaseModel):
    success: bool
    output: str
    error: str | None = None
    metadata: dict[str, JsonValue] = Field(default_factory=dict)


class AgentEngine(Protocol):
    def stream(self, request: AgentRunRequest, options: EngineOptions) -> AsyncIterator[AgentStreamEvent]: ...

    async def run(self, request: AgentRunRequest, options: EngineOptions) -> EngineRunResult: ...

File: deepgoal-ts/src/core/types/engine.ts
import type { AsyncIterableLike, JsonObject } from './common.js';
import { PermissionMode } from './enums.js';
import type { HookEvent, HookHandler } from './hooks.js';
import type { McpServerConfig, SkillConfig, ToolConfig } from './tools.js';

export interface EngineOptions {
  tools: ToolConfig[];
  mcpServers: Record<string, McpServerConfig>;
  skills: SkillConfig[];
  hooks: Partial<Record<HookEvent, HookHandler[]>>;
  permissionMode: PermissionMode;
}

export interface AgentRunRequest {
  prompt: string;
  cwd: string;
  metadata: JsonObject;
}

export type AgentStreamEventType = 'message' | 'tool_call' | 'tool_result' | 'error' | 'done';

export interface AgentStreamEvent {
  type: AgentStreamEventType;
  content?: string | undefined;
  metadata: JsonObject;
}

export interface EngineRunResult {
  success: boolean;
  output: string;
  error?: string | undefined;
  metadata: JsonObject;
}

export interface AgentEngine {
  stream(request: AgentRunRequest, options: EngineOptions): AsyncIterableLike<AgentStreamEvent>;
  run(request: AgentRunRequest, options: EngineOptions): Promise<EngineRunResult>;
}

export function createEngineOptions(options: Partial<EngineOptions> = {}): EngineOptions {
  return {
    tools: options.tools ?? [],
    mcpServers: options.mcpServers ?? {},
    skills: options.skills ?? [],
    hooks: options.hooks ?? {},
    permissionMode: options.permissionMode ?? PermissionMode.Default,
  };
}

export function createAgentRunRequest(prompt: string, cwd: string, metadata: JsonObject = {}): AgentRunRequest {
  return { prompt, cwd, metadata };
}

export function createAgentStreamEvent(type: AgentStreamEventType, content?: string, metadata: JsonObject = {}): AgentStreamEvent {
  return { type, content, metadata };
}

export function createEngineRunResult(success: boolean, output: string, metadata: JsonObject = {}, error?: string): EngineRunResult {
  return { success, output, metadata, error };
}
